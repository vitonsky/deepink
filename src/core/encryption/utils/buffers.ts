/**
 * Join buffers into a new buffer
 */
export const joinBuffers = (buffers: (ArrayBufferLike | Buffer | TypedArray)[]) => {
	const bufferLen = buffers.reduce((len, buffer) => len + buffer.byteLength, 0);
	const resultBuffer = new Uint8Array(bufferLen);

	let offset = 0;
	for (const buffer of buffers) {
		// We may receive a TypedArray instead of ArrayBuffer, so we have to get a buffer
		const rawBuffer = ArrayBuffer.isView(buffer) ? buffer.buffer : buffer;
		resultBuffer.set(new Uint8Array(rawBuffer), offset);
		offset += rawBuffer.byteLength;
	}

	return resultBuffer.buffer;
};

/**
 * Fill buffer with paddings to ensure buffer size multiple 16
 * Before use original data you have to remove padding
 */
export function fillBuffer(buffer: Uint8Array, blockSize = 16): [Uint8Array, number] {
	const padding = Math.ceil(buffer.length / blockSize) * blockSize - buffer.length;
	if (padding === 0) return [buffer, 0];

	// Create new buffer with padding
	const out = new Uint8Array(buffer.length + padding);
	out.set(buffer);

	return [out, padding];
}

/**
 * Buffer view to manipulate buffer slices
 */
export class BufferView {
	private readonly buffer;
	constructor(buffer: ArrayBuffer) {
		this.buffer = buffer;
	}

	public getBytes(offset: number = 0, end?: number) {
		return this.buffer.slice(offset, end);
	}

	public setBytes(buffer: ArrayBuffer, offset: number = 0) {
		if (offset > this.buffer.byteLength)
			throw new RangeError('Offset out of buffer size');
		if (offset + buffer.byteLength > this.buffer.byteLength)
			throw new RangeError(
				'Buffer size with current offset will out of buffer size',
			);

		const srcBufferView = new Uint8Array(buffer);
		const targetBufferView = new Uint8Array(this.buffer);
		for (let srcOffset = 0; srcOffset < buffer.byteLength; srcOffset++) {
			targetBufferView[offset + srcOffset] = srcBufferView[srcOffset];
		}
	}
}

type TypedArray =
	| Int8Array
	| Uint8Array
	| Uint8ClampedArray
	| Int16Array
	| Uint16Array
	| Int32Array
	| Uint32Array
	| Float32Array
	| Float64Array
	| BigInt64Array
	| BigUint64Array;

/**
 * Receive any buffer and returns a converted transferable buffer and codec for revert converting
 *
 * Useful to convert buffer for transfer and then convert received buffer back.
 *
 * @param buffer
 * @returns [buffer, convertor]
 */
export const convertBufferToTransferable = <
	T extends ArrayBufferLike | Buffer | TypedArray,
>(
	buffer: T,
): [ArrayBuffer, (buffer: ArrayBuffer) => T] => {
	if (globalThis.Buffer !== undefined && buffer instanceof Buffer) {
		return [
			buffer.buffer as ArrayBuffer,
			(buffer: ArrayBuffer) => {
				return Buffer.from(buffer) as T;
			},
		];
	}

	// Typed arrays
	const typedArray = [
		Int8Array,
		Uint8Array,
		Uint8ClampedArray,
		Int16Array,
		Uint16Array,
		Int32Array,
		Uint32Array,
		Float32Array,
		Float64Array,
		BigInt64Array,
		BigUint64Array,
	].find((proto) => buffer instanceof proto);
	if (typedArray && buffer instanceof typedArray) {
		return [
			(buffer as Uint8Array).buffer as ArrayBuffer,
			(buffer: ArrayBuffer) => {
				return new typedArray(buffer) as T;
			},
		];
	}

	return [
		buffer as ArrayBuffer,
		(buffer: ArrayBuffer) => {
			return buffer as T;
		},
	];
};
