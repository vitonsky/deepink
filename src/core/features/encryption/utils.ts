import { ENCRYPTION_ALGORITHM, ENCRYPTION_ALGORITHM_LIST } from './algorithms';

export const joinAlgorithms = (algorithms: ENCRYPTION_ALGORITHM[]) => {
	return algorithms.join('-');
};

export function isValidAlgorithm(algorithm: string): algorithm is ENCRYPTION_ALGORITHM {
	return Object.values(ENCRYPTION_ALGORITHM).includes(
		algorithm as ENCRYPTION_ALGORITHM,
	);
}

export function parseAlgorithms(algorithms: string) {
	return algorithms.split('-').map((name) => {
		if (!isValidAlgorithm(name)) {
			throw new Error(
				`Unsupported encryption algorithm: ${algorithms}. Supported: ${ENCRYPTION_ALGORITHM_LIST.join(
					', ',
				)}`,
			);
		}
		return name;
	});
}
