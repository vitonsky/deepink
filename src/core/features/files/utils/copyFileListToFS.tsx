import { joinPathSegments } from '@utils/fs/paths';

import { IFilesStorage } from '..';

export const copyFileListToFS = async (files: FileList, fs: IFilesStorage) => {
	// Read files
	for (const file of Array.from(files)) {
		let filePath = file.name;
		if (file.webkitRelativePath) {
			const relativePath = file.webkitRelativePath.split('/');
			filePath = joinPathSegments(
				relativePath.length > 1 ? relativePath.slice(1) : relativePath,
			);
		}

		const buffer = await file.arrayBuffer();
		await fs.write(filePath, buffer);
	}

	return fs;
};
