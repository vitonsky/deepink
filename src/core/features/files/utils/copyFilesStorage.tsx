import { IFilesStorage } from '@core/features/files';

// TODO: implement filter
/**
 * Copy all files from `source` FS to `target` FS and returns a `target`
 */
export const copyFilesStorage = async (source: IFilesStorage, target: IFilesStorage) => {
	for (const file of await source.list()) {
		const buffer = await source.get(file);
		if (!buffer) continue;

		await target.write(file, buffer);
	}

	return target;
};
