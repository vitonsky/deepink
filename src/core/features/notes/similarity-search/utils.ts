export function cosineSimilarity(a: number[], b: number[]) {
	if (a.length !== b.length) throw new Error('Vectors must be same length');
	let dot = 0;
	let normA = 0;
	let normB = 0;
	for (let i = 0; i < a.length; i++) {
		dot += a[i] * b[i];
		normA += a[i] * a[i];
		normB += b[i] * b[i];
	}
	return dot / (Math.sqrt(normA) * Math.sqrt(normB));
}
