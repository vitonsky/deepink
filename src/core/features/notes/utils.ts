import { INoteContent } from '@core/features/notes/index';

export const getNoteTitle = (note: INoteContent, limit?: number) => {
	let title = (note.title || note.text).trim();
	if (limit) {
		title = title.slice(0, limit).trim();
	}

	return title || 'Empty note';
};
