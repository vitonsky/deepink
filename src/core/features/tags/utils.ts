import { IResolvedTag } from '.';

// TODO: remove unused code
export const isTagsArray = (data: unknown): data is string[] =>
	Array.isArray(data) && data.every((item) => typeof item === 'string');

export const findParentTag = (resolvedTagName: string, tags: IResolvedTag[]) => {
	let parent: IResolvedTag | null = null;
	for (const tag of tags) {
		if (!resolvedTagName.startsWith(tag.resolvedName)) continue;

		// Remember only those parent who have more segments
		if (parent === null || parent.resolvedName.length < tag.resolvedName.length) {
			parent = tag;
		}
	}

	return parent;
};
