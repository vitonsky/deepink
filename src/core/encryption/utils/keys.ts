import { joinBuffers } from './buffers';

/**
 * Creates and return derived `CryptoKey` from a master password
 */
export async function getMasterKey(
	masterPassword: string | ArrayBuffer,
	salt: ArrayBuffer,
) {
	const codec = new TextEncoder();

	const key =
		typeof masterPassword === 'string'
			? codec.encode(masterPassword)
			: masterPassword;

	return await self.crypto.subtle
		.importKey('raw', key, { name: 'PBKDF2' }, false, ['deriveBits', 'deriveKey'])
		// We don't want to use key from original password, only derived keys,
		// so here we derive a new key based on password
		.then((key) =>
			self.crypto.subtle.deriveKey(
				{
					name: 'PBKDF2',
					salt: new Uint8Array(salt),
					iterations: 100000,
					hash: 'SHA-512',
				},
				key,
				{
					name: 'AES-GCM',
					length: 256,
				},

				// Whether or not the key is extractable
				true,

				// This usage limitations will be ignored, because we import and then export key with new permissions
				['encrypt', 'decrypt'],
			),
		)
		// Export and import of key is necessary, to bypass limitation about key can't be derived of derived key
		// Otherwise generates error `DOMException: key.algorithm does not match that of operation`
		.then((key) => self.crypto.subtle.exportKey('raw', key))
		.then((key) =>
			self.crypto.subtle.importKey('raw', key, { name: 'PBKDF2' }, false, [
				'deriveBits',
				'deriveKey',
			]),
		);
}

type KeyAlgorithm =
	| AlgorithmIdentifier
	| AesDerivedKeyParams
	| HmacImportParams
	| HkdfParams
	| Pbkdf2Params;

/**
 * Creates object to generate derived keys and bytes based on provided master key
 */
export async function getDerivedKeysManager(
	masterKey: CryptoKey,
	salt: Uint8Array<ArrayBuffer>,
) {
	const codec = new TextEncoder();

	const getScopedSalt = (scope: string) =>
		joinBuffers([codec.encode(scope).buffer, salt.buffer]);

	return {
		async getDerivedKey(context: string, algorithm: KeyAlgorithm) {
			return self.crypto.subtle.deriveKey(
				{
					name: 'PBKDF2',
					salt: getScopedSalt(context),
					iterations: 100000,
					hash: 'SHA-512',
				},
				masterKey,
				algorithm,

				// Whether or not the key is extractable (less secure) or not (more secure)
				// when false, the key can only be passed as a web crypto object, not inspected
				true,

				// this web crypto object will only be allowed for these functions
				['encrypt', 'decrypt'],
			);
		},

		async getDerivedBytes(context: string, length: number) {
			if (length % 8 !== 0) throw new TypeError('Length is not multiple to 8');

			return self.crypto.subtle.deriveBits(
				{
					name: 'PBKDF2',
					salt: getScopedSalt(context),
					iterations: 100000,
					hash: 'SHA-256',
				},
				masterKey,
				length,
			);
		},
	};
}
