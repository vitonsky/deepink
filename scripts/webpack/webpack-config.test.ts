import { getAppWindows } from './utils';

test('All windows and preload scripts must be found', () => {
	expect(getAppWindows()).toMatchSnapshot();
});
