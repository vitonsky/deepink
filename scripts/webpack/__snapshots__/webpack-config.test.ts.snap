// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`All windows and preload scripts must be found 1`] = `
[
  {
    "name": "about",
    "preloadScript": "src/windows/about/preload.ts",
    "renderer": "src/windows/about/renderer.tsx",
  },
  {
    "name": "main",
    "preloadScript": "src/windows/main/preload.ts",
    "renderer": "src/windows/main/renderer.tsx",
  },
]
`;
